external/js/cothority/src/network/connection.ts

```diff
@@ -242,3 +250,234 @@ export class LeaderConnection extends WebSocketConnection {
         super(roster.list[0].address, service);
     }
 }
+
+/**
+ * Nodes holds all nodes for all services in two lists - one active for the number of
+ * parallel open connections, and one reserve pool for connections that can take over if the
+ * active list fails.
+ *
+ * It does some advanced checking of which nodes to contact using the following steps:
+ * - sorting the nodes to contact by response time - fastest get contacted first
+ * - failing nodes are replaced by currently unused nodes
+ * - if a node is slower than the slowThreshold, the node is replaced with a node from the reserve pool
+ */
+export class Nodes {
+
+    /**
+     * Returns the number of nodes currently in the active queue.
+     */
+    get parallel(): number {
+        return this._parallel;
+    }
+
+    /**
+     * Resets the number of active connections
+     * @param p active connections
+     */
+    set parallel(p: number) {
+        if (Nodes.parallel > 0) {
+            p = Nodes.parallel;
+        }
+        if (p > this.addresses.length) {
+            this._parallel = this.addresses.length;
+        } else if (p > 0) {
+            this._parallel = p;
+        }
+        while (this._parallel > this.active.length) {
+            this.active.push(this.reserve.shift());
+        }
+        while (this._parallel < this.active.length) {
+            this.reserve.push(this.active.pop());
+        }
+    }
+    // This can be set from the outside to enforce a number of parallel requests. If it
+    // is > 0, it will be enforced.
+    static parallel: number = 0;
+    // the threshold for the ratio of response_node / response_fastest where the node is
+    // put in the 'reserve'.
+    static slowThreshold: number = 10;
+    private readonly addresses: string[];
+    // to the slowest node.
+    private readonly active: Node[] = [];
+    // is the one that was not used for the longest time.
+    private readonly reserve: Node[];
+    // Pool of available connections that are not in the active connections. The first node
+
+    // Current number of nodes in the active queue.
+    private _parallel: number;
+    // Holds the fastest connections for 'parallel' nodes. They are sorted from the fastest
+
+    constructor(r: Roster, parallel: number = 3) {
+        this.addresses = r.list.map((conode) => conode.getWebSocketAddress());
+        shuffle(this.addresses);
+        // Initialize the pool of connections
+        this.reserve = this.addresses.map((addr) => new Node(addr));
+        this.parallel = parallel;
+    }
+
+    /**
+     * Creates a new NodeList for one message.
+     * @param service
+     */
+    newList(service: string): NodeList {
+        if (Nodes.parallel > 0) {
+            this.parallel = Nodes.parallel;
+        }
```

Comment:
It's weird to change `parallel` is a seamingly unrelated call. This should have been handled by `set parallel`, no?

Reply:
There was a (bad) reason for that. I'll think it over again.